pkg/utils/controller/controller.go

```diff
@@ -33,25 +34,56 @@ var (
 	// There should be something wrong with the system and cannot be recovered by itself.
 	ErrUnexpectedBehavior = errors.New("unexpected behavior which cannot be handled by the controller")
 
+	// ErrExpectedBehavior indicates the current situation is expected, which can be recovered by itself after retries.
+	ErrExpectedBehavior = errors.New("expected behavior which can be recovered by itself")
```

Comment:
I am not 100% sure if I understand this part, and a little bit clarification would be greatly appreciated.

What would be if there's a resource version conflict (API error)? Is this an API error or an expected error? Similarly, what would be if an error is expected with sporadic occurrences and unexpected if repeated consistently?

Reply:
I think one way to interpret this is we can alert on too many ExpectedBehaviorError which will be treated the same as "UnExpectedError" while APIServerError will clearly say that there is something wrong with the api-server. We will look more into cluster status.

Reply:
For now there are api server errors, user errors and the rest are expected/unexpected logic errors. They should have no overlap.

agree with Ryan, the alert could be set based on the occurrences or status code returned by the API server.